File: fireplace-app/src/main/java/io/github/bric3/fireplace/ui/FrameResizeLabel.java
package io.github.bric3.fireplace.ui;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ComponentAdapter;
import java.awt.event.ComponentEvent;

public class FrameResizeLabel {
    private final JLabel dimensionLabel;
    private final JPanel dimensionOverlayPanel;
    private final Timer panelHider;

    public FrameResizeLabel() {
        dimensionLabel = new JLabel("hello");
        dimensionLabel.setVerticalAlignment(JLabel.CENTER);
        dimensionLabel.setHorizontalAlignment(JLabel.CENTER);
        dimensionLabel.setOpaque(true);
        dimensionLabel.setBorder(BorderFactory.createLineBorder(Color.black));
        dimensionOverlayPanel = new JPanel(new BorderLayout());
        dimensionOverlayPanel.add(dimensionLabel, BorderLayout.CENTER);
        dimensionOverlayPanel.setBackground(new Color(0, 0, 0, 0));
        dimensionOverlayPanel.setOpaque(false);
        dimensionOverlayPanel.setVisible(false);
        var textHeight = dimensionLabel.getFontMetrics(dimensionLabel.getFont()).getHeight();
        dimensionOverlayPanel.setMaximumSize(new Dimension(100, textHeight + 10));
        panelHider = new Timer(2_000, e -> dimensionOverlayPanel.setVisible(false));
        panelHider.setCoalesce(true);
    }

    public void installListener(JFrame frame) {
        frame.addComponentListener(new ComponentAdapter() {
            @Override
            public void componentResized(ComponentEvent e) {
                int height = frame.getHeight();
                int width = frame.getWidth();
                dimensionLabel.setText(height + " x " + width);
                dimensionOverlayPanel.setVisible(true);
                panelHider.restart();
            }
        });
    }

    public JComponent getComponent() {
        return dimensionOverlayPanel;
    }
}

File: fireplace-app/src/main/kotlin/io/github/bric3/fireplace/ui/FrameResizeLabel.kt
/*
 * Fireplace
 *
 * Copyright (c) 2021, Today - Brice Dutheil
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package io.github.bric3.fireplace.ui

import io.github.bric3.fireplace.core.ui.Colors
import io.github.bric3.fireplace.core.ui.LightDarkColor
import java.awt.BorderLayout
import java.awt.Color
import java.awt.Dimension
import java.awt.event.ComponentAdapter
import java.awt.event.ComponentEvent
import javax.swing.*

class FrameResizeLabel {
    private val dimensionLabel: JLabel = JLabel("hello").apply {
        verticalAlignment = JLabel.CENTER
        horizontalAlignment = JLabel.CENTER
        isOpaque = true
        border = BorderFactory.createLineBorder(
            LightDarkColor(
                Colors.panelForeground,
                Colors.panelBackground
            )
        )
    }
    private val dimensionOverlayPanel: JPanel = JPanel(BorderLayout()).apply {
        add(dimensionLabel, BorderLayout.CENTER)
        background = LightDarkColor(
            Colors.translucent_white_D0,
            Colors.translucent_black_80
        )
        isOpaque = false
        isVisible = false
    }
    private val panelHider: Timer

    init {
        dimensionOverlayPanel.maximumSize = dimensionLabel.getFontMetrics(dimensionLabel.font).let {
            val border = 10
            val textWidth = it.stringWidth("1000 x 1000") + 10
            val textHeight = it.height
            Dimension(textWidth + border, textHeight + border)
        }

        panelHider = Timer(2000) { _ -> dimensionOverlayPanel.isVisible = false }.apply {
            isCoalesce = true
        }
    }

    fun installListener(frame: JFrame) {
        frame.addComponentListener(object : ComponentAdapter() {
            override fun componentResized(e: ComponentEvent) {
                val height = frame.height
                val width = frame.width
                dimensionLabel.text = "$height x $width"
                dimensionOverlayPanel.isVisible = true
                panelHider.restart()
            }
        })
    }

    val component: JComponent
        get() = dimensionOverlayPanel
}
